From 607bb2f6d83648c7010487d3cfc48cf1a369b4a1 Mon Sep 17 00:00:00 2001
From: Sherief Farouk <sherief.personal@gmail.com>
Date: Fri, 7 Aug 2026 01:20:24 +0800
Subject: [PATCH] vkd3d: Implement ID3D12GraphicsCommandList9 and
 ID3D12Device10

Backport upstream commit 1317cf837bdcb83b94115134b33fd2da7661e7c6 onto the WineHua VKD3D-Proton 2.6 patch series.

The upstream swapchain-only macro updates are omitted because the 2.6 base still uses the older d3d12_swapchain_device_iface path and has no matching calls. The public COM interfaces and vtables are otherwise preserved.
---
 include/vkd3d_d3d12.idl            |  78 +++++++++++++++++++++
 libs/vkd3d/bundle.c                | 108 ++++++++++++++++++-----------
 libs/vkd3d/command.c               |  36 +++++++++-
 libs/vkd3d/command_list_profiled.h |  29 +++++++-
 libs/vkd3d/device.c                |  48 ++++++++++++-
 libs/vkd3d/device_profiled.h       |  35 +++++++++-
 libs/vkd3d/vkd3d_private.h         |  14 ++--
 7 files changed, 298 insertions(+), 50 deletions(-)

diff --git a/include/vkd3d_d3d12.idl b/include/vkd3d_d3d12.idl
index b154adc2..00cbeb91 100644
--- a/include/vkd3d_d3d12.idl
+++ b/include/vkd3d_d3d12.idl
@@ -3663,6 +3663,40 @@ interface ID3D12GraphicsCommandList6 : ID3D12GraphicsCommandList5
     void DispatchMesh(UINT x, UINT y, UINT z);
 }
 
+[
+    uuid(dd171223-8b61-4769-90e3-160ccde4e2c1),
+    object,
+    local,
+    pointer_default(unique)
+]
+interface ID3D12GraphicsCommandList7 : ID3D12GraphicsCommandList6
+{
+    void Barrier(UINT32 NumBarrierGroups, const void *pBarrierGroups);
+}
+
+[
+    uuid(ee936ef9-599d-4d28-938e-23c4ad05ce51),
+    object,
+    local,
+    pointer_default(unique)
+]
+interface ID3D12GraphicsCommandList8 : ID3D12GraphicsCommandList7
+{
+    void OMSetFrontAndBackStencilRef(UINT FrontStencilRef, UINT BackStencilRef);
+}
+
+[
+    uuid(34ed2808-ffe6-4c2b-b11a-cabd2b0c59e1),
+    object,
+    local,
+    pointer_default(unique)
+]
+interface ID3D12GraphicsCommandList9 : ID3D12GraphicsCommandList8
+{
+    void RSSetDepthBias(FLOAT DepthBias, FLOAT DepthBiasClamp, FLOAT SlopeScaledDepthBias);
+    void IASetIndexBufferStripCutValue(D3D12_INDEX_BUFFER_STRIP_CUT_VALUE IBStripCutValue);
+}
+
 typedef enum D3D12_TILE_RANGE_FLAGS
 {
     D3D12_TILE_RANGE_FLAG_NONE = 0x0,
@@ -4211,6 +4245,50 @@ interface ID3D12Device9 : ID3D12Device8
     HRESULT CreateCommandQueue1(const D3D12_COMMAND_QUEUE_DESC *desc,
             REFIID creator_id, REFIID riid, void **command_queue);
 }
+
+typedef enum D3D12_BARRIER_LAYOUT
+{
+    D3D12_BARRIER_LAYOUT_UNDEFINED = 0xFFFFFFFF
+} D3D12_BARRIER_LAYOUT;
+
+[
+    uuid(517f8718-aa66-49f9-b02b-a7ab89c06031),
+    object,
+    local,
+    pointer_default(unique)
+]
+interface ID3D12Device10 : ID3D12Device9
+{
+    HRESULT CreateCommittedResource3(const D3D12_HEAP_PROPERTIES *heap_properties,
+        D3D12_HEAP_FLAGS heap_flags,
+        const D3D12_RESOURCE_DESC1 *desc,
+        D3D12_BARRIER_LAYOUT initial_layout,
+        const D3D12_CLEAR_VALUE *optimized_clear_value,
+        ID3D12ProtectedResourceSession *protected_session,
+        UINT32 num_castable_formats,
+        DXGI_FORMAT *castable_formats,
+        REFIID iid,
+        void **resource);
+
+    HRESULT CreatePlacedResource2(ID3D12Heap *heap,
+        UINT64 heap_offset,
+        const D3D12_RESOURCE_DESC1 *desc,
+        D3D12_BARRIER_LAYOUT initial_layout,
+        const D3D12_CLEAR_VALUE *optimized_clear_value,
+        UINT32 num_castable_formats,
+        DXGI_FORMAT *castable_formats,
+        REFIID iid,
+        void **resource);
+
+    HRESULT CreateReservedResource2(const D3D12_RESOURCE_DESC *desc,
+        D3D12_BARRIER_LAYOUT initial_layout,
+        const D3D12_CLEAR_VALUE *optimized_clear_value,
+        ID3D12ProtectedResourceSession *protected_session,
+        UINT32 num_castable_formats,
+        DXGI_FORMAT *castable_formats,
+        REFIID iid,
+        void **resource);
+}
 [
     uuid(34ab647b-3cc8-46ac-841b-c0965645c046),
     object,
diff --git a/libs/vkd3d/bundle.c b/libs/vkd3d/bundle.c
index 48df4b63..dd45b906 100644
--- a/libs/vkd3d/bundle.c
+++ b/libs/vkd3d/bundle.c
@@ -262,12 +262,15 @@ static HRESULT STDMETHODCALLTYPE d3d12_bundle_QueryInterface(d3d12_command_list_
             || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList4)
             || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList5)
             || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList6)
+            || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList7)
+            || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList8)
+            || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList9)
             || IsEqualGUID(iid, &IID_ID3D12CommandList)
             || IsEqualGUID(iid, &IID_ID3D12DeviceChild)
             || IsEqualGUID(iid, &IID_ID3D12Object)
             || IsEqualGUID(iid, &IID_IUnknown))
     {
-        ID3D12GraphicsCommandList6_AddRef(iface);
+        ID3D12GraphicsCommandList9_AddRef(iface);
         *object = iface;
         return S_OK;
     }
@@ -430,7 +433,7 @@ static void d3d12_bundle_exec_draw_instanced(d3d12_command_list_iface *list, con
 {
     const struct d3d12_draw_instanced_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_DrawInstanced(list, args->vertex_count,
+    ID3D12GraphicsCommandList9_DrawInstanced(list, args->vertex_count,
             args->instance_count, args->first_vertex, args->first_instance);
 }
 
@@ -467,7 +470,7 @@ static void d3d12_bundle_exec_draw_indexed_instanced(d3d12_command_list_iface *l
 {
     const struct d3d12_draw_indexed_instanced_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_DrawIndexedInstanced(list, args->index_count,
+    ID3D12GraphicsCommandList9_DrawIndexedInstanced(list, args->index_count,
             args->instance_count, args->first_index, args->vertex_offset,
             args->first_instance);
 }
@@ -502,7 +505,7 @@ static void d3d12_bundle_exec_dispatch(d3d12_command_list_iface *list, const voi
 {
     const struct d3d12_dispatch_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_Dispatch(list, args->x, args->y, args->z);
+    ID3D12GraphicsCommandList9_Dispatch(list, args->x, args->y, args->z);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_Dispatch(d3d12_command_list_iface *iface,
@@ -570,7 +573,7 @@ static void d3d12_bundle_exec_ia_set_primitive_topology(d3d12_command_list_iface
 {
     const struct d3d12_ia_set_primitive_topology_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_IASetPrimitiveTopology(list, args->topology);
+    ID3D12GraphicsCommandList9_IASetPrimitiveTopology(list, args->topology);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_IASetPrimitiveTopology(d3d12_command_list_iface *iface,
@@ -607,7 +610,7 @@ static void d3d12_bundle_exec_om_set_blend_factor(d3d12_command_list_iface *list
 {
     const struct d3d12_om_set_blend_factor_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_OMSetBlendFactor(list, args->blend_factor);
+    ID3D12GraphicsCommandList9_OMSetBlendFactor(list, args->blend_factor);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_OMSetBlendFactor(d3d12_command_list_iface *iface,
@@ -635,7 +638,7 @@ static void d3d12_bundle_exec_om_set_stencil_ref(d3d12_command_list_iface *list,
 {
     const struct d3d12_om_set_stencil_ref_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_OMSetStencilRef(list, args->stencil_ref);
+    ID3D12GraphicsCommandList9_OMSetStencilRef(list, args->stencil_ref);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_OMSetStencilRef(d3d12_command_list_iface *iface,
@@ -660,7 +663,7 @@ static void d3d12_bundle_exec_set_pipeline_state(d3d12_command_list_iface *list,
 {
     const struct d3d12_set_pipeline_state_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetPipelineState(list, args->pipeline_state);
+    ID3D12GraphicsCommandList9_SetPipelineState(list, args->pipeline_state);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetPipelineState(d3d12_command_list_iface *iface,
@@ -704,7 +707,7 @@ static void d3d12_bundle_exec_set_compute_root_signature(d3d12_command_list_ifac
 {
     const struct d3d12_set_root_signature_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetComputeRootSignature(list, args->root_signature);
+    ID3D12GraphicsCommandList9_SetComputeRootSignature(list, args->root_signature);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetComputeRootSignature(d3d12_command_list_iface *iface,
@@ -723,7 +726,7 @@ static void d3d12_bundle_exec_set_graphics_root_signature(d3d12_command_list_ifa
 {
     const struct d3d12_set_root_signature_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetGraphicsRootSignature(list, args->root_signature);
+    ID3D12GraphicsCommandList9_SetGraphicsRootSignature(list, args->root_signature);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetGraphicsRootSignature(d3d12_command_list_iface *iface,
@@ -749,7 +752,7 @@ static void d3d12_bundle_exec_set_compute_root_descriptor_table(d3d12_command_li
 {
     const struct d3d12_set_root_descriptor_table_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetComputeRootDescriptorTable(list, args->parameter_index, args->base_descriptor);
+    ID3D12GraphicsCommandList9_SetComputeRootDescriptorTable(list, args->parameter_index, args->base_descriptor);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetComputeRootDescriptorTable(d3d12_command_list_iface *iface,
@@ -770,7 +773,7 @@ static void d3d12_bundle_exec_set_graphics_root_descriptor_table(d3d12_command_l
 {
     const struct d3d12_set_root_descriptor_table_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetGraphicsRootDescriptorTable(list, args->parameter_index, args->base_descriptor);
+    ID3D12GraphicsCommandList9_SetGraphicsRootDescriptorTable(list, args->parameter_index, args->base_descriptor);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetGraphicsRootDescriptorTable(d3d12_command_list_iface *iface,
@@ -799,7 +802,7 @@ static void d3d12_bundle_exec_set_compute_root_32bit_constant(d3d12_command_list
 {
     const struct d3d12_set_root_32bit_constant_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetComputeRoot32BitConstant(list, args->parameter_index, args->data, args->offset);
+    ID3D12GraphicsCommandList9_SetComputeRoot32BitConstant(list, args->parameter_index, args->data, args->offset);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetComputeRoot32BitConstant(d3d12_command_list_iface *iface,
@@ -821,7 +824,7 @@ static void d3d12_bundle_exec_set_graphics_root_32bit_constant(d3d12_command_lis
 {
     const struct d3d12_set_root_32bit_constant_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetGraphicsRoot32BitConstant(list, args->parameter_index, args->data, args->offset);
+    ID3D12GraphicsCommandList9_SetGraphicsRoot32BitConstant(list, args->parameter_index, args->data, args->offset);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetGraphicsRoot32BitConstant(d3d12_command_list_iface *iface,
@@ -852,7 +855,7 @@ static void d3d12_bundle_exec_set_compute_root_32bit_constants(d3d12_command_lis
 {
     const struct d3d12_set_root_32bit_constants_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetComputeRoot32BitConstants(list, args->parameter_index,
+    ID3D12GraphicsCommandList9_SetComputeRoot32BitConstants(list, args->parameter_index,
             args->constant_count, args->data, args->offset);
 }
 
@@ -880,7 +883,7 @@ static void d3d12_bundle_exec_set_graphics_root_32bit_constants(d3d12_command_li
 {
     const struct d3d12_set_root_32bit_constants_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetGraphicsRoot32BitConstants(list, args->parameter_index,
+    ID3D12GraphicsCommandList9_SetGraphicsRoot32BitConstants(list, args->parameter_index,
             args->constant_count, args->data, args->offset);
 }
 
@@ -915,7 +918,7 @@ static void d3d12_bundle_exec_set_compute_root_cbv(d3d12_command_list_iface *lis
 {
     const struct d3d12_set_root_descriptor_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetComputeRootConstantBufferView(list, args->parameter_index, args->address);
+    ID3D12GraphicsCommandList9_SetComputeRootConstantBufferView(list, args->parameter_index, args->address);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetComputeRootConstantBufferView(
@@ -936,7 +939,7 @@ static void d3d12_bundle_exec_set_graphics_root_cbv(d3d12_command_list_iface *li
 {
     const struct d3d12_set_root_descriptor_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetGraphicsRootConstantBufferView(list, args->parameter_index, args->address);
+    ID3D12GraphicsCommandList9_SetGraphicsRootConstantBufferView(list, args->parameter_index, args->address);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetGraphicsRootConstantBufferView(
@@ -957,7 +960,7 @@ static void d3d12_bundle_exec_set_compute_root_srv(d3d12_command_list_iface *lis
 {
     const struct d3d12_set_root_descriptor_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetComputeRootShaderResourceView(list, args->parameter_index, args->address);
+    ID3D12GraphicsCommandList9_SetComputeRootShaderResourceView(list, args->parameter_index, args->address);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetComputeRootShaderResourceView(
@@ -978,7 +981,7 @@ static void d3d12_bundle_exec_set_graphics_root_srv(d3d12_command_list_iface *li
 {
     const struct d3d12_set_root_descriptor_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetGraphicsRootShaderResourceView(list, args->parameter_index, args->address);
+    ID3D12GraphicsCommandList9_SetGraphicsRootShaderResourceView(list, args->parameter_index, args->address);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetGraphicsRootShaderResourceView(
@@ -999,7 +1002,7 @@ static void d3d12_bundle_exec_set_compute_root_uav(d3d12_command_list_iface *lis
 {
     const struct d3d12_set_root_descriptor_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetComputeRootUnorderedAccessView(list, args->parameter_index, args->address);
+    ID3D12GraphicsCommandList9_SetComputeRootUnorderedAccessView(list, args->parameter_index, args->address);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetComputeRootUnorderedAccessView(
@@ -1020,7 +1023,7 @@ static void d3d12_bundle_exec_set_graphics_root_uav(d3d12_command_list_iface *li
 {
     const struct d3d12_set_root_descriptor_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetGraphicsRootUnorderedAccessView(list, args->parameter_index, args->address);
+    ID3D12GraphicsCommandList9_SetGraphicsRootUnorderedAccessView(list, args->parameter_index, args->address);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetGraphicsRootUnorderedAccessView(
@@ -1045,14 +1048,14 @@ struct d3d12_ia_set_index_buffer_command
 
 static void d3d12_bundle_exec_ia_set_index_buffer_null(d3d12_command_list_iface *list, const void *args_v)
 {
-    ID3D12GraphicsCommandList6_IASetIndexBuffer(list, NULL);
+    ID3D12GraphicsCommandList9_IASetIndexBuffer(list, NULL);
 }
 
 static void d3d12_bundle_exec_ia_set_index_buffer(d3d12_command_list_iface *list, const void *args_v)
 {
     const struct d3d12_ia_set_index_buffer_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_IASetIndexBuffer(list, &args->view);
+    ID3D12GraphicsCommandList9_IASetIndexBuffer(list, &args->view);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_IASetIndexBuffer(d3d12_command_list_iface *iface,
@@ -1202,7 +1205,7 @@ static void d3d12_bundle_exec_set_marker(d3d12_command_list_iface *list, const v
 {
     const struct d3d12_debug_marker_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetMarker(list, args->metadata, args->data, args->data_size);
+    ID3D12GraphicsCommandList9_SetMarker(list, args->metadata, args->data, args->data_size);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetMarker(d3d12_command_list_iface *iface,
@@ -1223,7 +1226,7 @@ static void d3d12_bundle_exec_begin_event(d3d12_command_list_iface *list, const
 {
     const struct d3d12_debug_marker_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_BeginEvent(list, args->metadata, args->data, args->data_size);
+    ID3D12GraphicsCommandList9_BeginEvent(list, args->metadata, args->data, args->data_size);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_BeginEvent(d3d12_command_list_iface *iface,
@@ -1242,7 +1245,7 @@ static void STDMETHODCALLTYPE d3d12_bundle_BeginEvent(d3d12_command_list_iface *
 
 static void d3d12_bundle_exec_end_event(d3d12_command_list_iface *list, const void *args_v)
 {
-    ID3D12GraphicsCommandList6_EndEvent(list);
+    ID3D12GraphicsCommandList9_EndEvent(list);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_EndEvent(d3d12_command_list_iface *iface)
@@ -1269,7 +1272,7 @@ static void d3d12_bundle_exec_execute_indirect(d3d12_command_list_iface *list, c
 {
     const struct d3d12_execute_indirect_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_ExecuteIndirect(list, args->signature, args->max_count,
+    ID3D12GraphicsCommandList9_ExecuteIndirect(list, args->signature, args->max_count,
             args->arg_buffer, args->arg_offset, args->count_buffer, args->count_offset);
 }
 
@@ -1331,7 +1334,7 @@ static void d3d12_bundle_exec_om_set_depth_bounds(d3d12_command_list_iface *list
 {
     const struct d3d12_om_set_depth_bounds_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_OMSetDepthBounds(list, args->min, args->max);
+    ID3D12GraphicsCommandList9_OMSetDepthBounds(list, args->min, args->max);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_OMSetDepthBounds(d3d12_command_list_iface *iface,
@@ -1360,7 +1363,7 @@ static void d3d12_bundle_exec_set_sample_positions(d3d12_command_list_iface *lis
     const struct d3d12_set_sample_positions_command *args = args_v;
 
     /* The sample position array is non-const but does not get written to */
-    ID3D12GraphicsCommandList6_SetSamplePositions(list, args->sample_count,
+    ID3D12GraphicsCommandList9_SetSamplePositions(list, args->sample_count,
             args->pixel_count, (D3D12_SAMPLE_POSITION*)args->positions);
 }
 
@@ -1403,7 +1406,7 @@ static void d3d12_bundle_exec_set_view_instance_mask(d3d12_command_list_iface *l
 {
     const struct d3d12_set_view_instance_mask_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetViewInstanceMask(list, args->mask);
+    ID3D12GraphicsCommandList9_SetViewInstanceMask(list, args->mask);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetViewInstanceMask(d3d12_command_list_iface *iface, UINT mask)
@@ -1429,7 +1432,7 @@ static void d3d12_bundle_exec_write_buffer_immediate(d3d12_command_list_iface *l
 {
     const struct d3d12_write_buffer_immediate_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_WriteBufferImmediate(list, args->count, args->parameters, args->modes);
+    ID3D12GraphicsCommandList9_WriteBufferImmediate(list, args->count, args->parameters, args->modes);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_WriteBufferImmediate(d3d12_command_list_iface *iface,
@@ -1525,7 +1528,7 @@ static void d3d12_bundle_exec_set_pipeline_state1(d3d12_command_list_iface *list
 {
     const struct d3d12_set_pipeline_state1_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_SetPipelineState1(list, args->state_object);
+    ID3D12GraphicsCommandList9_SetPipelineState1(list, args->state_object);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_SetPipelineState1(d3d12_command_list_iface *iface,
@@ -1550,7 +1553,7 @@ static void d3d12_bundle_exec_dispatch_rays(d3d12_command_list_iface *list, cons
 {
     const struct d3d12_dispatch_rays_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_DispatchRays(list, &args->desc);
+    ID3D12GraphicsCommandList9_DispatchRays(list, &args->desc);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_DispatchRays(d3d12_command_list_iface *iface,
@@ -1576,14 +1579,14 @@ static void d3d12_bundle_exec_rs_set_shading_rate(d3d12_command_list_iface *list
 {
     const struct d3d12_rs_set_shading_rate_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_RSSetShadingRate(list, args->base, args->combiners);
+    ID3D12GraphicsCommandList9_RSSetShadingRate(list, args->base, args->combiners);
 }
 
 static void d3d12_bundle_exec_rs_set_shading_rate_base(d3d12_command_list_iface *list, const void *args_v)
 {
     const struct d3d12_rs_set_shading_rate_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_RSSetShadingRate(list, args->base, NULL);
+    ID3D12GraphicsCommandList9_RSSetShadingRate(list, args->base, NULL);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_RSSetShadingRate(d3d12_command_list_iface *iface,
@@ -1613,7 +1616,7 @@ static void d3d12_bundle_exec_rs_set_shading_rate_image(d3d12_command_list_iface
 {
     const struct d3d12_rs_set_shading_rate_image_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_RSSetShadingRateImage(list, args->image);
+    ID3D12GraphicsCommandList9_RSSetShadingRateImage(list, args->image);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_RSSetShadingRateImage(d3d12_command_list_iface *iface,
@@ -1632,7 +1635,7 @@ static void d3d12_bundle_exec_dispatch_mesh(d3d12_command_list_iface *list, cons
 {
     const struct d3d12_dispatch_command *args = args_v;
 
-    ID3D12GraphicsCommandList6_DispatchMesh(list, args->x, args->y, args->z);
+    ID3D12GraphicsCommandList9_DispatchMesh(list, args->x, args->y, args->z);
 }
 
 static void STDMETHODCALLTYPE d3d12_bundle_DispatchMesh(d3d12_command_list_iface *iface, UINT x, UINT y, UINT z)
@@ -1648,7 +1651,27 @@ static void STDMETHODCALLTYPE d3d12_bundle_DispatchMesh(d3d12_command_list_iface
     args->z = z;
 }
 
-static CONST_VTBL struct ID3D12GraphicsCommandList6Vtbl d3d12_bundle_vtbl =
+static void STDMETHODCALLTYPE d3d12_bundle_Barrier(d3d12_command_list_iface *iface, UINT32 NumBarrierGroups, const void *pBarrierGroups)
+{
+    WARN("iface %p, NumBarrierGroups %u, D3D12_BARRIER_GROUP %p ignored!\n", iface, NumBarrierGroups, pBarrierGroups);
+}
+
+static void STDMETHODCALLTYPE d3d12_bundle_OMSetFrontAndBackStencilRef(d3d12_command_list_iface *iface, UINT FrontStencilRef, UINT BackStencilRef)
+{
+    WARN("iface %p, FrontStencilRef %u, BackStencilRef %u ignored!\n", iface, FrontStencilRef, BackStencilRef);
+}
+
+static void STDMETHODCALLTYPE d3d12_bundle_RSSetDepthBias(d3d12_command_list_iface *iface, FLOAT DepthBias, FLOAT DepthBiasClamp, FLOAT SlopeScaledDepthBias)
+{
+    WARN("iface %p, DepthBias %f, DepthBiasClamp %f, SlopeScaledDepthBias %f ignored!\n", iface, DepthBias, DepthBiasClamp, SlopeScaledDepthBias);
+}
+
+static void STDMETHODCALLTYPE d3d12_bundle_IASetIndexBufferStripCutValue(d3d12_command_list_iface *iface, D3D12_INDEX_BUFFER_STRIP_CUT_VALUE IBStripCutValue)
+{
+    WARN("iface %p, IBStripCutValue %u ignored!\n", iface, IBStripCutValue);
+}
+
+static CONST_VTBL struct ID3D12GraphicsCommandList9Vtbl d3d12_bundle_vtbl =
 {
     /* IUnknown methods */
     d3d12_bundle_QueryInterface,
@@ -1741,6 +1764,13 @@ static CONST_VTBL struct ID3D12GraphicsCommandList6Vtbl d3d12_bundle_vtbl =
     d3d12_bundle_RSSetShadingRateImage,
     /* ID3D12GraphicsCommandList6 methods */
     d3d12_bundle_DispatchMesh,
+    /* ID3D12GraphicsCommandList7 methods */
+    d3d12_bundle_Barrier,
+    /* ID3D12GraphicsCommandList8 methods */
+    d3d12_bundle_OMSetFrontAndBackStencilRef,
+    /* ID3D12GraphicsCommandList9 methods */
+    d3d12_bundle_RSSetDepthBias,
+    d3d12_bundle_IASetIndexBufferStripCutValue,
 };
 
 HRESULT d3d12_bundle_create(struct d3d12_device *device,
diff --git a/libs/vkd3d/command.c b/libs/vkd3d/command.c
index 12f2b0d5..8e4708e2 100644
--- a/libs/vkd3d/command.c
+++ b/libs/vkd3d/command.c
@@ -3992,6 +3992,9 @@ HRESULT STDMETHODCALLTYPE d3d12_command_list_QueryInterface(d3d12_command_list_i
             || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList4)
             || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList5)
             || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList6)
+            || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList7)
+            || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList8)
+            || IsEqualGUID(iid, &IID_ID3D12GraphicsCommandList9)
             || IsEqualGUID(iid, &IID_ID3D12CommandList)
             || IsEqualGUID(iid, &IID_ID3D12DeviceChild)
             || IsEqualGUID(iid, &IID_ID3D12Object)
@@ -9932,7 +9935,31 @@ static void STDMETHODCALLTYPE d3d12_command_list_DispatchMesh(d3d12_command_list
     FIXME("iface %p, x %u, y %u, z %u stub!", iface, x, y, z);
 }
 
-static CONST_VTBL struct ID3D12GraphicsCommandList6Vtbl d3d12_command_list_vtbl =
+static void STDMETHODCALLTYPE d3d12_command_list_Barrier(d3d12_command_list_iface *iface, UINT32 NumBarrierGroups, const void *pBarrierGroups)
+{
+    FIXME("iface %p, NumBarrierGroups %u, D3D12_BARRIER_GROUP %p stub!\n",
+        iface, NumBarrierGroups, pBarrierGroups);
+}
+
+static void STDMETHODCALLTYPE d3d12_command_list_OMSetFrontAndBackStencilRef(d3d12_command_list_iface *iface, UINT FrontStencilRef, UINT BackStencilRef)
+{
+    FIXME("iface %p, FrontStencilRef %u, BackStencilRef %u stub!\n",
+        iface, FrontStencilRef, BackStencilRef);
+}
+
+static void STDMETHODCALLTYPE d3d12_command_list_RSSetDepthBias(d3d12_command_list_iface *iface, FLOAT DepthBias, FLOAT DepthBiasClamp, FLOAT SlopeScaledDepthBias)
+{
+    FIXME("iface %p, DepthBias %f, DepthBiasClamp %f, SlopeScaledDepthBias %f stub!\n",
+        iface, DepthBias, DepthBiasClamp, SlopeScaledDepthBias);
+}
+
+static void STDMETHODCALLTYPE d3d12_command_list_IASetIndexBufferStripCutValue(d3d12_command_list_iface *iface, D3D12_INDEX_BUFFER_STRIP_CUT_VALUE IBStripCutValue)
+{
+    FIXME("iface %p, IBStripCutValue %u stub!\n",
+        iface, IBStripCutValue);
+}
+
+static CONST_VTBL struct ID3D12GraphicsCommandList9Vtbl d3d12_command_list_vtbl =
 {
     /* IUnknown methods */
     d3d12_command_list_QueryInterface,
@@ -10025,6 +10052,13 @@ static CONST_VTBL struct ID3D12GraphicsCommandList6Vtbl d3d12_command_list_vtbl
     d3d12_command_list_RSSetShadingRateImage,
     /* ID3D12GraphicsCommandList6 methods */
     d3d12_command_list_DispatchMesh,
+    /* ID3D12GraphicsCommandList7 methods */
+    d3d12_command_list_Barrier,
+    /* ID3D12GraphicsCommandList8 methods */
+    d3d12_command_list_OMSetFrontAndBackStencilRef,
+    /* ID3D12GraphicsCommandList9 methods */
+    d3d12_command_list_RSSetDepthBias,
+    d3d12_command_list_IASetIndexBufferStripCutValue,
 };
 
 #ifdef VKD3D_ENABLE_PROFILING
diff --git a/libs/vkd3d/command_list_profiled.h b/libs/vkd3d/command_list_profiled.h
index 3d74b15f..0acbfd28 100644
--- a/libs/vkd3d/command_list_profiled.h
+++ b/libs/vkd3d/command_list_profiled.h
@@ -464,7 +464,27 @@ static void STDMETHODCALLTYPE d3d12_command_list_DispatchMesh_profiled(d3d12_com
     COMMAND_LIST_PROFILED_CALL(DispatchMesh, iface, x, y, z);
 }
 
-static CONST_VTBL struct ID3D12GraphicsCommandList6Vtbl d3d12_command_list_vtbl_profiled =
+static void STDMETHODCALLTYPE d3d12_command_list_Barrier_profiled(d3d12_command_list_iface *iface, UINT32 NumBarrierGroups, const void *pBarrierGroups)
+{
+    COMMAND_LIST_PROFILED_CALL(iface, Barrier, NumBarrierGroups, pBarrierGroups);
+}
+
+static void STDMETHODCALLTYPE d3d12_command_list_OMSetFrontAndBackStencilRef_profiled(d3d12_command_list_iface *iface, UINT FrontStencilRef, UINT BackStencilRef)
+{
+    COMMAND_LIST_PROFILED_CALL(iface, OMSetFrontAndBackStencilRef, FrontStencilRef, BackStencilRef);
+}
+
+static void STDMETHODCALLTYPE d3d12_command_list_RSSetDepthBias_profiled(d3d12_command_list_iface *iface, FLOAT DepthBias, FLOAT DepthBiasClamp, FLOAT SlopeScaledDepthBias)
+{
+    COMMAND_LIST_PROFILED_CALL(iface, RSSetDepthBias, DepthBias, DepthBiasClamp, SlopeScaledDepthBias);
+}
+
+static void STDMETHODCALLTYPE d3d12_command_list_IASetIndexBufferStripCutValue_profiled(d3d12_command_list_iface *iface, D3D12_INDEX_BUFFER_STRIP_CUT_VALUE IBStripCutValue)
+{
+    COMMAND_LIST_PROFILED_CALL(iface, IBStripCutValue);
+}
+
+static CONST_VTBL struct ID3D12GraphicsCommandList9Vtbl d3d12_command_list_vtbl_profiled =
 {
     /* IUnknown methods */
     d3d12_command_list_QueryInterface,
@@ -557,6 +577,13 @@ static CONST_VTBL struct ID3D12GraphicsCommandList6Vtbl d3d12_command_list_vtbl_
     d3d12_command_list_RSSetShadingRateImage_profiled,
     /* ID3D12GraphicsCommandList6 methods */
     d3d12_command_list_DispatchMesh_profiled,
+    /* ID3D12GraphicsCommandList7 methods */
+    d3d12_command_list_Barrier_profiled,
+    /* ID3D12GraphicsCommandList8 methods */
+    d3d12_command_list_OMSetFrontAndBackStencilRef_profiled,
+    /* ID3D12GraphicsCommandList9 methods */
+    d3d12_command_list_RSSetDepthBias_profiled,
+    d3d12_command_list_IASetIndexBufferStripCutValue_profiled,
 };
 
 #endif
diff --git a/libs/vkd3d/device.c b/libs/vkd3d/device.c
index 1af99ff4..58287394 100644
--- a/libs/vkd3d/device.c
+++ b/libs/vkd3d/device.c
@@ -2648,6 +2648,7 @@ HRESULT STDMETHODCALLTYPE d3d12_device_QueryInterface(d3d12_device_iface *iface,
             || IsEqualGUID(riid, &IID_ID3D12Device7)
             || IsEqualGUID(riid, &IID_ID3D12Device8)
             || IsEqualGUID(riid, &IID_ID3D12Device9)
+            || IsEqualGUID(riid, &IID_ID3D12Device10)
             || IsEqualGUID(riid, &IID_ID3D12Object)
             || IsEqualGUID(riid, &IID_IUnknown))
     {
@@ -5192,7 +5193,48 @@ static HRESULT STDMETHODCALLTYPE d3d12_device_CreateCommandQueue1(d3d12_device_i
     return d3d12_device_CreateCommandQueue(iface, desc, iid, command_queue);
 }
 
-CONST_VTBL struct ID3D12Device9Vtbl d3d12_device_vtbl =
+static HRESULT STDMETHODCALLTYPE d3d12_device_CreateCommittedResource3(d3d12_device_iface *iface,
+    const D3D12_HEAP_PROPERTIES *heap_properties, D3D12_HEAP_FLAGS heap_flags,
+    const D3D12_RESOURCE_DESC1 *desc, D3D12_BARRIER_LAYOUT initial_layout,
+    const D3D12_CLEAR_VALUE *optimized_clear_value, ID3D12ProtectedResourceSession *protected_session,
+    UINT32 num_castable_formats, DXGI_FORMAT *castable_formats, REFIID iid, void **resource)
+{
+    FIXME("iface %p, heap_properties %p, heap_flags %u, desc %p, initial_layout %u, "
+            "optimized_clear_value %p, protected_session %p, num_castable_formats %u, "
+            "castable_formats %p, iid %s, resource %p stub!\n", iface,
+            heap_properties, heap_flags, desc, initial_layout, optimized_clear_value,
+            protected_session, num_castable_formats, castable_formats, debugstr_guid(iid), resource);
+
+    return E_NOTIMPL;
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_device_CreatePlacedResource2(d3d12_device_iface *iface,
+    ID3D12Heap *heap, UINT64 heap_offset, const D3D12_RESOURCE_DESC1 *desc, D3D12_BARRIER_LAYOUT initial_layout,
+    const D3D12_CLEAR_VALUE *optimized_clear_value, UINT32 num_castable_formats,
+    DXGI_FORMAT *castable_formats, REFIID iid, void **resource)
+{
+    FIXME("iface %p, heap %p, heap_offset %#"PRIx64", desc %p, initial_layout %u, optimized_clear_value %p, "
+            "num_castable_formats %u, castable_formats %p, iid %s, resource %p stub!\n", iface,
+            heap_offset, desc, initial_layout, optimized_clear_value, num_castable_formats,
+            castable_formats, debugstr_guid(iid), resource);
+
+    return E_NOTIMPL;
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_device_CreateReservedResource2(d3d12_device_iface *iface,
+    const D3D12_RESOURCE_DESC *desc, D3D12_BARRIER_LAYOUT initial_layout, const D3D12_CLEAR_VALUE *optimized_clear_value,
+    ID3D12ProtectedResourceSession *protected_session, UINT32 num_castable_formats,
+    DXGI_FORMAT *castable_formats, REFIID iid, void **resource)
+{
+    FIXME("iface %p, desc %p initial_layout %u, optimized_clear_value %p, protected_session %p, "
+            "num_castable_formats %u, castable_formats %p, iid %s, resource %p stub!\n", iface,
+            initial_layout, optimized_clear_value, protected_session, num_castable_formats,
+            castable_formats, debugstr_guid(iid), resource);
+
+    return E_NOTIMPL;
+}
+
+CONST_VTBL struct ID3D12Device10Vtbl d3d12_device_vtbl =
 {
     /* IUnknown methods */
     d3d12_device_QueryInterface,
@@ -5282,6 +5324,10 @@ CONST_VTBL struct ID3D12Device9Vtbl d3d12_device_vtbl =
     d3d12_device_CreateShaderCacheSession,
     d3d12_device_ShaderCacheControl,
     d3d12_device_CreateCommandQueue1,
+    /* ID3D12Device10 methods */
+    d3d12_device_CreateCommittedResource3,
+    d3d12_device_CreatePlacedResource2,
+    d3d12_device_CreateReservedResource2,
 };
 
 #ifdef VKD3D_ENABLE_PROFILING
diff --git a/libs/vkd3d/device_profiled.h b/libs/vkd3d/device_profiled.h
index d1a66cdd..181f35b5 100644
--- a/libs/vkd3d/device_profiled.h
+++ b/libs/vkd3d/device_profiled.h
@@ -239,7 +239,36 @@ static void STDMETHODCALLTYPE d3d12_device_CreateSamplerFeedbackUnorderedAccessV
     DEVICE_PROFILED_CALL(CreateSamplerFeedbackUnorderedAccessView, iface, target_resource, feedback_resource, descriptor);
 }
 
-CONST_VTBL struct ID3D12Device9Vtbl d3d12_device_vtbl_profiled =
+static HRESULT STDMETHODCALLTYPE d3d12_device_CreateCommittedResource3_profiled(d3d12_device_iface *iface,
+    const D3D12_HEAP_PROPERTIES *heap_properties, D3D12_HEAP_FLAGS heap_flags,
+    const D3D12_RESOURCE_DESC1 *desc, D3D12_BARRIER_LAYOUT initial_layout,
+    const D3D12_CLEAR_VALUE *optimized_clear_value, ID3D12ProtectedResourceSession *protected_session,
+    UINT32 num_castable_formats, DXGI_FORMAT *castable_formats, REFIID iid, void **resource)
+{
+    DEVICE_PROFILED_CALL(CreateCommittedResource3, iface, heap_properties, heap_flags, desc, initial_layout,
+                            optimized_clear_value, protected_session, num_castable_formats, castable_formats,
+                            iid, resource);
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_device_CreatePlacedResource2_profiled(d3d12_device_iface *iface,
+    UINT64 heap_offset, const D3D12_RESOURCE_DESC1 *desc, D3D12_BARRIER_LAYOUT initial_layout,
+    const D3D12_CLEAR_VALUE *optimized_clear_value, UINT32 num_castable_formats,
+    DXGI_FORMAT *castable_formats, REFIID iid, void **resource)
+{
+    DEVICE_PROFILED_CALL(CreatePlacedResource2, iface, heap_offset, desc, initial_layout, optimized_clear_value,
+                            num_castable_formats, castable_formats, iid, resource);
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_device_CreateReservedResource2_profiled(d3d12_device_iface *iface,
+    D3D12_BARRIER_LAYOUT initial_layout, const D3D12_CLEAR_VALUE *optimized_clear_value,
+    ID3D12ProtectedResourceSession *protected_session, UINT32 num_castable_formats,
+    DXGI_FORMAT *castable_formats, REFIID iid, void **resource)
+{
+    DEVICE_PROFILED_CALL(CreateReservedResource2, iface, initial_layout, optimized_clear_value, protected_session,
+                            num_castable_formats, castable_formats, iid, resource);
+}
+
+CONST_VTBL struct ID3D12Device10Vtbl d3d12_device_vtbl_profiled =
 {
     /* IUnknown methods */
     d3d12_device_QueryInterface,
@@ -329,6 +358,10 @@ CONST_VTBL struct ID3D12Device9Vtbl d3d12_device_vtbl_profiled =
     d3d12_device_CreateShaderCacheSession,
     d3d12_device_ShaderCacheControl,
     d3d12_device_CreateCommandQueue1,
+    /* ID3D12Device10 methods */
+    d3d12_device_CreateCommittedResource3_profiled,
+    d3d12_device_CreatePlacedResource2_profiled,
+    d3d12_device_CreateReservedResource2_profiled,
 };
 
 #endif
diff --git a/libs/vkd3d/vkd3d_private.h b/libs/vkd3d/vkd3d_private.h
index c39de9d5..91f41829 100644
--- a/libs/vkd3d/vkd3d_private.h
+++ b/libs/vkd3d/vkd3d_private.h
@@ -1916,7 +1916,7 @@ struct vkd3d_dynamic_state
 };
 
 /* ID3D12CommandList */
-typedef ID3D12GraphicsCommandList6 d3d12_command_list_iface;
+typedef ID3D12GraphicsCommandList9 d3d12_command_list_iface;
 
 enum vkd3d_initial_transition_type
 {
@@ -2920,7 +2920,7 @@ struct vkd3d_cached_command_allocator
 };
 
 /* ID3D12Device */
-typedef ID3D12Device9 d3d12_device_iface;
+typedef ID3D12Device10 d3d12_device_iface;
 
 struct vkd3d_descriptor_qa_global_info;
 struct vkd3d_descriptor_qa_heap_buffer_data;
@@ -3015,9 +3015,9 @@ static inline struct d3d12_device *unsafe_impl_from_ID3D12Device(d3d12_device_if
 
 static inline struct d3d12_device *impl_from_ID3D12Device(d3d12_device_iface *iface)
 {
-    extern CONST_VTBL struct ID3D12Device9Vtbl d3d12_device_vtbl;
+    extern CONST_VTBL struct ID3D12Device10Vtbl d3d12_device_vtbl;
 #ifdef VKD3D_ENABLE_PROFILING
-    extern CONST_VTBL struct ID3D12Device9Vtbl d3d12_device_vtbl_profiled;
+    extern CONST_VTBL struct ID3D12Device10Vtbl d3d12_device_vtbl_profiled;
 #endif
     if (!iface)
         return NULL;
@@ -3069,17 +3069,17 @@ static inline const struct vkd3d_memory_info_domain *d3d12_device_get_memory_inf
 
 static inline HRESULT d3d12_device_query_interface(struct d3d12_device *device, REFIID iid, void **object)
 {
-    return ID3D12Device9_QueryInterface(&device->ID3D12Device_iface, iid, object);
+    return ID3D12Device10_QueryInterface(&device->ID3D12Device_iface, iid, object);
 }
 
 static inline ULONG d3d12_device_add_ref(struct d3d12_device *device)
 {
-    return ID3D12Device9_AddRef(&device->ID3D12Device_iface);
+    return ID3D12Device10_AddRef(&device->ID3D12Device_iface);
 }
 
 static inline ULONG d3d12_device_release(struct d3d12_device *device)
 {
-    return ID3D12Device9_Release(&device->ID3D12Device_iface);
+    return ID3D12Device10_Release(&device->ID3D12Device_iface);
 }
 
 static inline unsigned int d3d12_device_get_descriptor_handle_increment_size(
-- 
2.43.0

